File: golang/noob/once/wrapper.go
package main

import (
	"comwrapper"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"unsafe"

	"git.iflytek.com/AIaaS/xsf/utils"
)

var (
	wLogger  *utils.Logger
	logLevel = "debug"
	logCount = 10
	logSize  = 30
	logAsync = true
	logPath  = "/log/app/wrapper/wrapper.log"
)

var traceLogFunc func(usrTag string, key string, value string) (code int)

// WrapperInit 插件初始化, 全局只调用一次. 本地调试时, cfg参数由aiges.toml提供
func WrapperInit(cfg map[string]string) (err error) {
	fmt.Println("---- wrapper init ----")
	for k, v := range cfg {
		fmt.Printf("config param %s=%s\n", k, v)
	}
	if v, ok := cfg["log_path"]; ok {
		logPath = v
	}
	if v, ok := cfg["log_level"]; ok {
		logLevel = v
	}
	if v, ok := cfg["log_count"]; ok {
		logCount, _ = strconv.Atoi(v)
	}
	if v, ok := cfg["log_size"]; ok {
		logSize, _ = strconv.Atoi(v)
	}
	if cfg["log_async"] == "false" {
		logAsync = false
	}

	wLogger, err = utils.NewLocalLog(
		utils.SetAsync(logAsync),
		utils.SetLevel(logLevel),
		utils.SetFileName(logPath),
		utils.SetMaxSize(logSize),
		utils.SetMaxBackups(logCount),
	)
	if err != nil {
		return fmt.Errorf("loggerErr:%v", err)
	}

	wLogger.Debugw("WrapperInit successful")
	return
}

// WrapperExec 非流式调用. params参数由客户端请求时传入, 本地调试时, params参数由xtest.toml提供
func WrapperExec(usrTag string, params map[string]string, reqData []comwrapper.WrapperData) (respData []comwrapper.WrapperData, err error) {
	sid := params["sid"]
	wLogger.Debugw("WrapperExec Enter", "params", params, "sid", sid)

	respData = make([]comwrapper.WrapperData, 0, len(reqData))
	for _, req := range reqData {
		text := strings.TrimSpace(string(req.Data))
		numbers := strings.Split(text, ",")
		var total int64
		for _, number := range numbers {
			n, err := strconv.ParseInt(number, 10, 64)
			if err != nil {
				wLogger.Errorw("WrapperExec params validation", "error", err.Error(), "sid", sid)
				return nil, errors.New("support integer input only")
			}
			total += n
		}

		// 上报trace日志, 建议上传关键、小体积的日志信息 或 不上报
		if code := traceLogFunc(usrTag, "total", strconv.Itoa(int(total))); code != 0 {
			wLogger.Errorw("WrapperExec trace total", "code", code, "sid", sid)
		}

		respData = append(respData, comwrapper.WrapperData{
			Key:      "result",
			Data:     []byte(strconv.Itoa(int(total))),
			Desc:     nil,
			Encoding: "",
			Type:     comwrapper.DataText,
			Status:   comwrapper.DataOnce,
		})
	}

	return
}

// WrapperDestroy 插件资源销毁
func WrapperDestroy(hdl interface{}) (err error) {
	return
}

// WrapperCreate 插件会话实例创建, 每次建立会话请求时调用. 本地调试时, params参数由xtest.toml提供
func WrapperCreate(usrTag string, params map[string]string, prsIds []int, cb comwrapper.CallBackPtr) (hdl unsafe.Pointer, err error) {
	return
}

// WrapperWrite 数据写入
func WrapperWrite(hdl unsafe.Pointer, req []comwrapper.WrapperData) (err error) {
	return
}

// WrapperRead 数据结果读取
func WrapperRead(hdl unsafe.Pointer) (respData []comwrapper.WrapperData, err error) {
	return
}

func WrapperFini() (err error) {
	return
}

func WrapperVersion() (version string) {
	return
}
func WrapperLoadRes(res comwrapper.WrapperData, resId int) (err error) {
	return
}
func WrapperUnloadRes(resId int) (err error) {
	return
}
func WrapperDebugInfo(hdl interface{}) (debug string) {
	return
}

func WrapperSetCtrl(fType comwrapper.CustomFuncType, f interface{}) (err error) {
	switch fType {
	case comwrapper.FuncTraceLog:
		cf := f.(func(usrTag string, key string, value string) (code int))
		traceLogFunc = cf
		fmt.Println("WrapperSetCtrl traceLogFunc set successful.")
	default:

	}
	return
}

File: cpp/noob/once/wrapper.cpp
#include "aiges/wrapper.h"
#include <curl/curl.h> // 头文件路径：/usr/include/x86_64-linux-gnu   库文件路径：/usr/lib/x86_64-linux-gnu
#include <cstring>
#include <iostream>
#include "aiges/wrapper.h"
#include "wrapper_error.h"
#include "nlohmann/json.hpp"
#include "spdlog/sinks/rotating_file_sink.h"
#include "spdlog/spdlog.h"

const char *version = "1.0.0";

std::map<std::string, spdlog::level::level_enum> log_level_map = {
    {"trace", spdlog::level::trace},
    {"debug", spdlog::level::debug},
    {"info", spdlog::level::info},
    {"warn", spdlog::level::warn},
    {"err", spdlog::level::err},
    {"critical", spdlog::level::critical},
    {"off", spdlog::level::off},
};

std::string log_level = "debug";
std::string log_path = "/log/app/wrapper.log";
int log_size = 10;  // 日志文件大小, 默认10MB
int log_count = 30; // 日志文件留存个数, 默认30个
std::string resp_key = "result";

std::string getLogDirectory(const std::string &logPath)
{
  // 查找最后一个斜杠的位置
  size_t lastSlashPos = logPath.find_last_of("/\\");

  // 提取目录路径
  if (lastSlashPos != std::string::npos)
  {
    return logPath.substr(0, lastSlashPos);
  }
  else
  {
    return ""; // 如果路径中不包含斜杠，返回空字符串
  }
}

void initlog()
{
  std::string command = "mkdir -p " + getLogDirectory(log_path);
  auto ret = system(command.c_str());
  if (ret != 0)
  {
    spdlog::error("initlog create log directory failed. ret:{}", ret);
  }

  // 设置全局日志级别为 debug
  spdlog::set_level(log_level_map[log_level]);

  // 设置日志格式
  spdlog::set_pattern("[%l] [%Y-%m-%d %H:%M:%S.%e] [%t] %v");

  SPDLOG_TRACE("Some trace message with param {}", {});
  SPDLOG_DEBUG("Some debug message");

  auto file_logger = spdlog::rotating_logger_mt("noob-wrapper", log_path, 1048576 * log_size, log_count, true);
  spdlog::set_default_logger(file_logger);

  // 根据需要决定是否设置日志刷新方式
  spdlog::flush_on(spdlog::level::debug);
}

wrapperMeterCustom meterPtr;
wrapperTraceLog traceLogPtr;
WrapperAPI int wrapperSetCtrl(CtrlType type, void *func)
{
  if (type == CTMeterCustom)
  {
    spdlog::debug("wrapperSetCtrl type:CTMeterCustom");
    meterPtr = (wrapperMeterCustom)func;
  }
  if (type == CTTraceLog)
  {
    spdlog::debug("wrapperSetCtrl type:CTTraceLog");
    traceLogPtr = (wrapperTraceLog)func;
  }
  return 0;
}

WrapperAPI int wrapperInit(pConfig cfg)
{
  // 传入配置
  while (cfg != nullptr)
  {
    if (cfg->key != nullptr && cfg->value != nullptr)
    {
      if (strcmp(cfg->key, "log_level") == 0)
      {
        log_level = std::string(cfg->value);
      }
      if (strcmp(cfg->key, "log_path") == 0)
      {
        log_path = std::string(cfg->value);
      }
      if (strcmp(cfg->key, "log_size") == 0)
      {
        log_size = std::stoi(std::string(cfg->value));
      }
      if (strcmp(cfg->key, "log_count") == 0)
      {
        log_count = std::stoi(std::string(cfg->value));
      }
    }
    cfg = cfg->next;
  }

  initlog();

  spdlog::info("wrapperInit log_level: {}", log_level);
  spdlog::info("wrapperInit log_path: {}", log_path);
  spdlog::info("wrapperInit log_size: {}", log_size);
  spdlog::info("wrapperInit log_count: {}", log_count);

  spdlog::debug("wrapperInit success");
  return Success;
}

int WrapperAPI wrapperExec(const char *usrTag, pParamList params, pDataList reqData, pDataList *respData, unsigned int psrIds[], int psrCnt)
{
  spdlog::debug("wrapperExec start...");
  std::string sid;

  // 获取请求参数：先取出 sid
  pParamList tmpParams = params;
  while (tmpParams != nullptr)
  {
    if (strcmp(tmpParams->key, "sid") == 0)
    {
      sid = std::string(tmpParams->value);
    }
    tmpParams = tmpParams->next;
  }

  // 获取文本
  if (reqData->len == 0)
  {
    spdlog::debug("wrapperExec req data empty, sid:{}", sid);
    return ReqDataEmpty;
  }
  std::string text = std::string((char *)reqData->data, reqData->len);
  spdlog::debug("wrapperExec input text: {} sid: {}", text, sid);

  // json 结果适配：
  nlohmann::json js;
  js["data"] = text;
  js["length"] = reqData->len;

  std::string result = js.dump();
  spdlog::debug("wrapperExec iflyResult:{} sid:{}", result, sid);

  // 封装响应
  *respData = new DataList();

  char *dynamicKey = new char[resp_key.size() + 1];
  strcpy(dynamicKey, resp_key.c_str());
  (*respData)->key = dynamicKey;

  char *dynamicData = new char[result.size() + 1];
  strcpy(dynamicData, result.c_str());
  (*respData)->data = (void *)dynamicData;

  (*respData)->len = result.size();
  (*respData)->desc = nullptr;
  (*respData)->next = nullptr;
  (*respData)->type = DataText;
  (*respData)->status = DataOnce;

  spdlog::debug("wrapperExec finish..., sid:{}", sid);

  return Success;
}

int WrapperAPI wrapperExecFree(const char *usrTag, pDataList *respData)
{ // respData 二级指针
  spdlog::debug("wrapperExecFree start...");
  // std::cout << "wrapperExecFree start..." << std::endl;
  pDataList resultPtr = *respData;
  if (resultPtr != nullptr)
  {
    if (resultPtr->key != nullptr)
    {
      delete[] resultPtr->key;
      resultPtr->key = nullptr;
    }
    if (resultPtr->data != nullptr)
    {
      delete[] static_cast<char *>(resultPtr->data);
      resultPtr->data = nullptr;
    }
    delete resultPtr;
    resultPtr = nullptr;
  }
  spdlog::debug("wrapperExecFree finish...");
  return Success;
}

WrapperAPI int wrapperFini()
{
  return Success;
}

WrapperAPI const char *wrapperVersion() { return version; }

WrapperAPI const char *wrapperError(int err) { return ErrorToString(err); }

WrapperAPI const void *wrapperCreate(const char *usrTag, pParamList params, wrapperCallback cb, unsigned int psrIds[], int psrCnt, int *errNum)
{
  return nullptr;
}

WrapperAPI int wrapperWrite(const void *handle, pDataList reqData)
{
  return NotSupportError;
}

WrapperAPI int wrapperRead(const void *handle, pDataList *respData)
{
  return NotSupportError;
}

WrapperAPI int wrapperDestroy(const void *handle)
{
  return NotSupportError;
}

WrapperAPI const char *wrapperDebugInfo(const void *handle)
{
  return "not support right now";
}

WrapperAPI int wrapperLoadRes(pDataList perData, unsigned int resId)
{
  return NotSupportError;
}

WrapperAPI int wrapperUnloadRes(unsigned int resId)
{
  return NotSupportError;
}

int WrapperAPI wrapperExecAsync(const char *usrTag, pParamList params, pDataList reqData, wrapperCallback callback, int timeout, unsigned int psrIds[], int psrCnt)
{
  return NotSupportError;
}
